model/flow/protocol_state.go

```diff
@@ -90,7 +105,22 @@ func (c *EpochStateContainer) Copy() *EpochStateContainer {
 		SetupID:          c.SetupID,
 		CommitID:         c.CommitID,
 		ActiveIdentities: c.ActiveIdentities.Copy(),
+		EpochExtensions:  clone.Clone(c.EpochExtensions),
+	}
+}
+
+// ExtendEpoch appends an epoch extension to the EpochStateContainer.
+// It is expected that the epoch extension is contiguous with the last extension, otherwise an exception is returned.
+// No errors are expected during normal operation.
+func (c *EpochStateContainer) ExtendEpoch(epochExtension EpochExtension) error {
+	if len(c.EpochExtensions) > 0 {
+		lastExtension := c.EpochExtensions[len(c.EpochExtensions)-1]
+		if lastExtension.FinalView+1 != epochExtension.FirstView {
+			return fmt.Errorf("epoch extension is not contiguous with the last extension")
+		}
 	}
+	c.EpochExtensions = append(c.EpochExtensions, epochExtension)
+	return nil
 }
```

Comment:
I am not entirely convinced about moving stuff into `RichProtocolStateEntry`. 

> Furthermore, we were thinking about: In case of entering EFT when a future epoch is in its setup phase, we want to drop the next epoch's setup data from the state, because this data is anyway outdated.
If we dropped the next epoch's setup data, we would need to modify not only the EpochStateContainer but also the RichProtocolStateEntry.

I am not concerned about this since `RichProtocolStateEntry` enriches existing data in the `ProtocolStateEntry`, state machine operates on `ProtocolStateEntry` when applying updates and `RichProtocolStateEntry` is created only when reading it from DB when the block has been applied which leads us to a valid representation even if we have removed next epoch.

> While EpochStateContainer cannot check the consistency of the views for the very first extension, RichProtocolStateEntry can! As far as I can tell, this would not break anything, because all the higher-level logic anyway works on the RichProtocolStateEntry.

It is true, I agree. Based on your previous [comment](https://github.com/onflow/flow-go/pull/5773/files#r1583884183) I am inclined to move the logic into state machine all together and localize changes to the state container in one place. 

Let me know if you want to discuss it further.

Reply:
[`51cf99d` (#5773)](https://github.com/onflow/flow-go/pull/5773/commits/51cf99d1924d545a6cd5d9f754fd35420608767d)